src/sessions/VirtualPageLoadTimer.ts

```diff
@@ -0,0 +1,245 @@
+import { Config } from '../orchestration/Orchestration';
+import { RecordEvent } from '../plugins/Plugin';
+import { PageManager, Page } from '../sessions/PageManager';
+
+type Fetch = typeof fetch;
+type Send = () => void;
+
+/**
+ * Maintains the core logic for virtual page load timing functionality.
+ * (1) Holds all virtual page load timing related resources
+ * (2) Intercepts outgoing XMLHttpRequests and Fetch requests and listens for DOM changes
+ * (3) Records virtual page load
+ */
+export class VirtualPageLoadTimer extends MonkeyPatched {
+    /** Unique ID of virtual page load periodic checker. */
+    private periodicCheckerId;
+    /** Unique ID of virtual page load timeout checker. */
+    private timeoutCheckerId;
+    /** Observer to liten for DOM changes. */
+    private domMutationObserver: MutationObserver;
+    /** Set to hold outgoing XMLHttpRequests for current virtual page. */
+    private ongoingRequests: Set<XMLHttpRequest>;
+    /** Buffer to hold outgoing XMLHttpRequests before new page is created. */
+    private requestBuffer: Set<XMLHttpRequest>;
+    /** Indicate the number of active Fetch requests. */
+    private fetchCounter: number;
+    /** Indicate the status of the current Page */
+    private isPageLoaded: boolean;
+    /** Indicate the current page's load end time value. */
+    private latestEndTime: number;
+
+    private config: Config;
+    private pageManager: PageManager;
+    private record: RecordEvent;
+
+    constructor(pageManager: PageManager, config: Config, record: RecordEvent) {
+        super();
+        this.periodicCheckerId = undefined;
+        this.timeoutCheckerId = undefined;
+        this.domMutationObserver = new MutationObserver(this.resetInterval);
+        this.ongoingRequests = new Set<XMLHttpRequest>();
+        this.requestBuffer = new Set();
+        this.fetchCounter = 0;
+        this.isPageLoaded = true;
+        this.latestEndTime = 0;
+
+        this.config = config;
+        this.pageManager = pageManager;
+        this.record = record;
+    }
+
+    /** Initializes timing related resources for current page. */
+    public startTiming() {
+        // Clean up existing timer objects and mutationObserver
+        if (this.periodicCheckerId) {
+            clearInterval(this.periodicCheckerId);
+        }
+        if (this.timeoutCheckerId) {
+            clearTimeout(this.timeoutCheckerId);
+        }
+        this.domMutationObserver.disconnect();
+
+        // Initialize timer objects and start observing
+        this.periodicCheckerId = setInterval(
+            this.checkLoadStatus,
+            this.config.routeChangeLoadCheckInterval
+        );
+        this.timeoutCheckerId = setTimeout(
+            this.declareTimeout,
+            this.config.routeChangeTimeout
+        );
+        this.domMutationObserver.observe(document, {
+            subtree: true,
+            childList: true,
+            attributes: true,
+            characterData: true
```

Comment:
question: Would it be sufficient to observe added/deleted nodes? I.e., and ignore attributes and character data?

Reply:
For first iteration, we will remove unnecessary attributes and only observe added/deleted nodes.